datasources/whatsapp/locations.go

```diff
@@ -0,0 +1,38 @@
+package whatsapp
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/timelinize/timelinize/timeline"
+)
+
+func extractLocation(content []string) (string, timeline.Metadata, bool) {
+	if len(content) < 2 {
+		return "", nil, false
+	}
+
+	trimmedContent := strings.TrimSpace(content[1])
+	meta := make(timeline.Metadata)
+	if idx := strings.Index(trimmedContent, foursquarePrefix); idx != -1 {
+		meta["Location Foursquare ID"] = trimmedContent[idx+len(foursquarePrefix):]
+	} else if idx := strings.Index(trimmedContent, googleMapsPrefix); idx != -1 {
+		var lat float64
+		var lng float64
+		if _, err := fmt.Sscanf(trimmedContent[idx+len(googleMapsPrefix):], "%f,%f", &lat, &lng); err != nil {
+			return "", nil, false
+		}
+
+		meta["Location Latitude"] = lat
+		meta["Location Longitude"] = lng
```

Comment:
```suggestion
		meta["Latitude"] = lat
		meta["Longitude"] = lng
```

How would you feel about this?

It's a bit unconventional, I suppose, since normally you'd put this in the item's geocoords directly; but that would imply the item was at that location. Is that the case of these messages? In one sense, it is, since the message is about that location, even if it is not describing a person being at that location. So, I'm open to discussing this now that I see it.

Reply:
I was wondering about exactly this!

I think it'd be worth being clear in your data model that this location is "over there", rather than "over here". (These are sometimes a 'current location', but the likelihood is vanishingly small, as the current location messages are auto deleted after their fixed timespan of ~1 hour — and few people will be taking an export in that hour.) I'd worry that putting them without a namespace might make them easier to misunderstand them as being about "where the sender is"?

The second reason I went with a namespace is the "named location" pins are Foursquare IDs, which are no particular use right now, so they're also under the same namespace of `Location`.

I'm beginning to think you might want structured data for your Metadata! Or at least an optionally structured part to it, so we can add specific context to certain fields (like polls and locations!)

With post-processors that pull out data from messages I'd definitely want to do things like extract all the restaurants I've ever posted about (or visited!), or collect WebArchive copies of every website I posted about.

Even making a prettier UI for polls would mean agreement between front end and back on where the Poll metadata is stored. Agreement between two parts is an API; and an unstructured map as an API is asking for trouble!

What do you think? I'd probably put this in a `ReferencedLocation` _named_ key if this was in place.

Reply:
Thanks!

So, I'm currently thinking that the website docs will be for end users, but we can use the project wiki for developers. What do you think of that plan? Both will be collaborative.

Reply:
Ahh yes of course — that sounds perfect; if you add a metadata-themed wiki page in the right place I'll add some context on these new keys!